add 9/2 sprint functional (WebSocket)

Olga-Preob opened the pull request.

Olga-Preob committed:
add 9/2 sprint functional (WebSocket)

gennady-bars commented on src/services/middleware/socket-middleware.js:
```diff
@@ -0,0 +1,78 @@
+  wsOnClose: WS_ON_CLOSE
+}
+
+export function socketMiddleware(wsActions) {
```
Хорошо сделали `socketMiddleware`

gennady-bars commented on src/pages/login/login.jsx:
```diff
@@ -14,7 +14,8 @@ import styles from './login.module.css';
+  const isRequest = useSelector((store) => store.userReducer.isRequest);
+  const isFailed = useSelector((store) => store.userReducer.isFailed);
 
   const [emailValue, setEmailValue] = useState('');
   const [passwordValue, setPasswordValue] = useState('');
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues={}) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -3,51 +3,65 @@ import { useDispatch, useSelector } from 'react-redux';
+  }, [isAuthChecked, dispatch]);
+
+  useEffect(() => {
+    ingredientsArr.length === 0 && dispatch(fetchGetIngredients());
```
`dispatch(fetchGetIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/components/order-list/order-list.jsx:
```diff
@@ -0,0 +1,163 @@
+  const orderNumber = currentOrderNumber ? currentOrderNumber : number;
+
+  useEffect(() => {
+    ingredientsArr.length === 0 && dispatch(fetchGetIngredients());
```
`dispatch(fetchGetIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/components/user-info/orders-history/orders-history.jsx:
```diff
@@ -1,10 +1,123 @@
+  const wsError = useSelector((store) => store.socketReducer.wsError);
+
+  useEffect(() => {
+    !ingredientsArr.length && dispatch(fetchGetIngredients());
```
`dispatch(fetchGetIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/pages/feed/feed.jsx:
```diff
@@ -1,11 +1,110 @@
+  const wsError = useSelector((store) => store.socketReducer.wsError);
+
+  useEffect(() => {
+    !ingredientsArr.length && dispatch(fetchGetIngredients());
```
`dispatch(fetchGetIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/pages/home/home.jsx:
```diff
@@ -12,35 +13,48 @@ import styles from './home.module.css';
-      dispatch(fetchGetIngredients());
-    }
-  }, [itemsSuccess, dispatch]);
+    ingredientsArr.length === 0 && dispatch(fetchGetIngredients());
```
`dispatch(fetchGetIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/pages/ingredient/ingredient.jsx:
```diff
@@ -14,40 +14,40 @@ function Ingredient() {
-    if (!itemsSuccess) {
-      dispatch(fetchGetIngredients());
-    }
+    ingredientsArr.length === 0 && dispatch(fetchGetIngredients());
```
`dispatch(fetchGetIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/pages/order/order.jsx:
```diff
@@ -0,0 +1,125 @@
+  const isFailed = useSelector((store) => store.orderInteractionReducer.isFailed);
+
+  useEffect(() => {
+    ingredientsArr.length === 0 && dispatch(fetchGetIngredients());
```
`dispatch(fetchGetIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/components/user-info/orders-history/orders-history.jsx:
```diff
@@ -1,10 +1,123 @@
+    !wsConnected && dispatch(wsConnectFeedOnToken());
+
+    return () => {
+      wsConnected && dispatch(disconnect());
```
Отлично, что закрываете соединение сокетов при покидании страниц

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -1,34 +1,31 @@
-        type: RESET_ORDER
-      })
-    );
+    typeOfModal === 'ingredientInfo' && navigate(-1, {replace: true});
+    typeOfModal === 'orderInfo' && navigate(-1, {replace: true});
   }
```
```js

    typeOfModal === 'ingredientInfo' && navigate(-1, {replace: true});
    typeOfModal === 'orderInfo' && navigate(-1, {replace: true});
```


Это нужно удалить из `Modal`

Внутри компонента `Modal` не должно быть **хардкода** строк и компонентов. 

`Modal` является **универсальным** компонентом для любых модальных окон. Он не должен знать внутри себя какие-то **конкретные** данные (экшены) для конкретных попапов.

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -40,7 +37,8 @@ function Modal({ children }) {
         });
 
-        typeOfModal === 'ingredient' && navigate(-1);
+        typeOfModal === 'ingredientInfo' && navigate(-1, {replace: true});
+        typeOfModal === 'orderInfo' && navigate(-1, {replace: true});
```
`Modal` является **универсальным** компонентом для любых модальных окон. Он не должен знать внутри себя какие-то **конкретные** данные (экшены) для конкретных попапов.

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -49,26 +47,33 @@ function Modal({ children }) {
-              {title}
-            </p>
+
+            {typeOfModal === 'orderInfo' && (
```
`Modal` является **универсальным** компонентом для любых модальных окон. Он не должен знать внутри себя какие-то **конкретные** данные (экшены) для конкретных попапов.

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -49,26 +47,33 @@ function Modal({ children }) {
+              </p>
+            )}
+
+            {typeOfModal !== 'orderInfo' && (
```
`Modal` является **универсальным** компонентом для любых модальных окон. Он не должен знать внутри себя какие-то **конкретные** данные (экшены) для конкретных попапов.

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:




* Хорошо сделали `socketMiddleware`
* Отличная логика попапов с заказами
* Отлично, что закрываете соединение сокетов при покидании страниц 



но есть некоторые недочеты:






* `dispatch(fetchGetIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.
* `Modal` является **универсальным** компонентом для любых модальных окон. Он не должен знать внутри себя какие-то **конкретные** данные (экшены) для конкретных попапов.

Olga-Preob committed:
fix: fix the logic of modal windows,
remove needles dispatch(fetchGetIngredients()),
fix work with inputs (add custom hook for inputs)

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -37,15 +38,14 @@ function Modal({ title = '', children }) {
 
-        typeOfModal === 'ingredientInfo' && navigate(-1, {replace: true});
-        typeOfModal === 'orderInfo' && navigate(-1, {replace: true});
+        navigateState.isNavigate && navigate(navigateState.to, { replace: navigateState.replace });
```
У Вас есть функция `handleOnClose` для закрытия, чтобы не дублировать этот код

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились. Красивый проект получится уже!

Удачного дальнейшего обучения.

Olga-Preob committed:
fix: remove code duplication in modal.jsx,
correct a typo in the folder name for the 404 page

Olga-Preob committed:
docs: update readme

Olga-Preob committed:
docs: update readme

Olga-Preob committed:
docs: update readme

Olga-Preob committed:
docs: update readme

Olga-Preob committed:
docs: update readme

Olga-Preob committed:
docs: update readme

Olga-Preob committed:
docs: update readme

Olga-Preob committed:
docs: update readme

Olga-Preob committed:
docs: update readme

Olga-Preob committed:
docs: update readme

Olga-Preob committed:
docs: update readme
